File: benchmark/punctuation-restoration/parity/koen-punct.mjs
// JS side of the koen_punctuation parity check (see parity/koen-punct.py for the reference side).
//
// usage:
//   node parity/koen-punct.mjs run <fp32|int8|int8-matmul>   replay results/parity-koen-punct-ref-torch.json rows through
//                                                             models/koen-punct-core.mjs on that graph (WASM, 1 thread):
//                                                             tokenizer ids, per-token labels, output text, and
//                                                             session.run latency at 64/128/256 content tokens
//   node parity/koen-punct.mjs report                        compare every results/parity-koen-punct-{js,ref}-*.json present
import { readFile, writeFile, access } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { ort, Tokenizer, fileReader } from '../lib/node-env.mjs';
import { loadEngine, LOCAL_DIR, WINDOW } from '../models/koen-punct-core.mjs';
import { analyze } from '../lib/text.mjs';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const results = (name) => join(root, 'results', name);
const readJson = async (p) => JSON.parse(await readFile(p, 'utf8'));
const exists = (p) => access(p).then(() => true, () => false);
const FILES = { fp32: 'model.onnx', int8: 'model.int8.onnx', 'int8-matmul': 'model.int8-matmul.onnx' };
const sameArr = (a, b) => a.length === b.length && a.every((x, i) => x === b[i]);

async function run(variant) {
  const modelFile = FILES[variant];
  if (!modelFile) throw new Error(`unknown variant ${variant}`);
  ort.env.wasm.numThreads = 1;
  const ref = await readJson(results('parity-koen-punct-ref-torch.json'));
  const t0 = performance.now();
  const engine = await loadEngine({ ort, Tokenizer, readFile: fileReader(LOCAL_DIR), executionProviders: ['wasm'] }, modelFile);
  const loadMs = performance.now() - t0;
  console.log(`${variant}: loaded ${modelFile} in ${loadMs.toFixed(0)} ms`);

  const tok = { rows: 0, idMismatch: [], unkRows: [] };
  const rows = [];
  for (const r of ref.rows) {
    tok.rows++;
    const ids = engine.encode(r.input, true).ids;
    if (!sameArr(ids, r.ids)) tok.idMismatch.push({ id: r.id, variant: r.variant, js: ids, py: r.ids });
    if (ids.includes(3)) tok.unkRows.push(`${r.id}/${r.variant}`);
    const s = performance.now();
    const p = await engine.predict(r.input);
    const out = await engine.run(r.input);
    rows.push({
      id: r.id, variant: r.variant, lang: r.lang, labels: p.labels, windows: p.windows,
      punct: out.text, mismatches: out.mismatches, ms: Math.round((performance.now() - s) * 10) / 10,
    });
  }
  console.log(`tokenizer: ${tok.rows} rows, id mismatches ${tok.idMismatch.length}, rows with <unk> ${tok.unkRows.length}`);
  console.log(`character-mapping misses: ${rows.reduce((a, r) => a + r.mismatches, 0)}`);

  // session.run latency on single windows of exactly N content tokens (+<s></s>), median of 15 after 3 warm-ups.
  const bench = {};
  for (const lang of ['ko', 'en']) {
    const long = ref.rows.find((r) => r.id === `${lang}-long`);
    const content = long.ids.slice(1, -1);
    for (const n of [64, 128, 256]) {
      const ids = Array.from({ length: n }, (_, i) => content[i % content.length]);
      for (let i = 0; i < 3; i++) await engine.runWindow(ids);
      const ts = [];
      for (let i = 0; i < 15; i++) {
        const s = performance.now();
        await engine.runWindow(ids);
        ts.push(performance.now() - s);
      }
      ts.sort((a, b) => a - b);
      bench[`${lang}/${n}`] = Math.round(ts[7] * 10) / 10;
    }
  }
  console.log('session.run median ms (1 thread):', bench);
  await engine.release();
  const dst = results(`parity-koen-punct-js-${variant}.json`);
  await writeFile(dst, JSON.stringify({ variant, modelFile, loadMs, tokenizer: tok, bench, rows }, null, 1));
  console.log(`wrote ${dst}`);
}

/** Scored (skeleton position, mark class) pairs of one output, as lib/text.mjs scores them. */
const events = (s) => new Set([...analyze(s).marks].map(([k, c]) => `${k}:${c}`));

/** Content-token labels of a row: JS rows store them without <s></s>, reference rows with. */
const contentLabels = (r) => (r.ids ? r.labels.slice(1, -1) : r.labels);

function compare(label, a, b, { filter = () => true, limit = 10 } = {}) {
  const byKey = new Map(b.map((r) => [`${r.id}/${r.variant}`, r]));
  const c = { label, rows: 0, same: 0, labelsSame: 0, labelsTotal: 0, rowsLabelsEqual: 0, eventsAgree: 0, eventsUnion: 0, diffs: [] };
  for (const r of a) {
    const o = byKey.get(`${r.id}/${r.variant}`);
    if (!o || !filter(r)) continue;
    c.rows++;
    if (r.punct === o.punct) c.same++;
    else if (c.diffs.length < limit) c.diffs.push({ row: `${r.id}/${r.variant}`, a: r.punct, b: o.punct });
    const la = contentLabels(r);
    const lb = contentLabels(o);
    if (sameArr(la, lb)) c.rowsLabelsEqual++;
    for (let i = 0; i < Math.min(la.length, lb.length); i++) if (la[i] === lb[i]) c.labelsSame++;
    c.labelsTotal += Math.max(la.length, lb.length);
    const ea = events(r.punct);
    const eb = events(o.punct);
    for (const e of ea) if (eb.has(e)) c.eventsAgree++;
    c.eventsUnion += new Set([...ea, ...eb]).size;
  }
  const pctOf = (x, y) => (y ? ((100 * x) / y).toFixed(1) : '-');
  console.log(`${label.padEnd(52)} rows ${c.same}/${c.rows} (${pctOf(c.same, c.rows)}%)  label-rows ${c.rowsLabelsEqual}/${c.rows}  tokens ${c.labelsSame}/${c.labelsTotal} (${pctOf(c.labelsSame, c.labelsTotal)}%)  marks ${c.eventsAgree}/${c.eventsUnion} (${pctOf(c.eventsAgree, c.eventsUnion)}%)`);
  return c;
}

async function report() {
  const load = async (name) => ((await exists(results(name))) ? (await readJson(results(name))).rows : null);
  const torch = await load('parity-koen-punct-ref-torch.json');
  const py = {};
  const js = {};
  for (const v of Object.keys(FILES)) {
    py[v] = await load(`parity-koen-punct-ref-${v}.json`);
    js[v] = await load(`parity-koen-punct-js-${v}.json`);
  }
  const scopes = [
    ['ko corpus', (r) => r.lang === 'ko' && r.variant !== 'long'],
    ['en corpus', (r) => r.lang === 'en' && r.variant !== 'long'],
    ['long rows', (r) => r.variant === 'long'],
  ];
  const out = [];
  const both = (label, a, b) => {
    if (!a || !b) return;
    for (const [scope, filter] of scopes) out.push(compare(`${label} [${scope}]`, a, b, { filter }));
  };
  for (const v of Object.keys(FILES)) both(`JS ${v} vs torch`, js[v], torch);
  for (const v of Object.keys(FILES)) both(`onnxruntime-py ${v} vs torch`, py[v], torch);
  for (const v of ['int8', 'int8-matmul']) both(`JS ${v} vs JS fp32`, js[v], js.fp32);
  for (const v of Object.keys(FILES)) both(`JS ${v} vs onnxruntime-py ${v}`, js[v], py[v]);
  for (const c of out) {
    if (!c.diffs.length) continue;
    console.log(`\n-- ${c.label}: first ${c.diffs.length} diffs`);
    for (const d of c.diffs) console.log(`   ${d.row}\n     a: ${JSON.stringify(d.a)}\n     b: ${JSON.stringify(d.b)}`);
  }
  const meta = {};
  for (const v of Object.keys(FILES)) {
    const p = results(`parity-koen-punct-js-${v}.json`);
    if (!(await exists(p))) continue;
    const j = await readJson(p);
    meta[v] = {
      loadMs: Math.round(j.loadMs), bench: j.bench,
      tokenizer: { rows: j.tokenizer.rows, idMismatch: j.tokenizer.idMismatch.length, unkRows: j.tokenizer.unkRows },
      mappingMisses: j.rows.reduce((a, r) => a + r.mismatches, 0),
      maxWindows: Math.max(...j.rows.map((r) => r.windows)),
    };
  }
  const longTokens = torch?.filter((r) => r.variant === 'long').map((r) => `${r.id}: ${r.n_tokens} tokens (window ${WINDOW})`);
  console.log('\n', JSON.stringify({ meta, longTokens }, null, 1));
  await writeFile(results('parity-koen-punct.json'), JSON.stringify({ comparisons: out, js: meta, longTokens }, null, 1));
  console.log(`wrote ${results('parity-koen-punct.json')}`);
}

const [cmd, arg] = process.argv.slice(2);
if (cmd === 'run') await run(arg);
else if (cmd === 'report') await report();
else throw new Error('usage: node parity/koen-punct.mjs run <fp32|int8|int8-matmul> | report');

File: benchmark/punctuation-restoration/parity/koen-punct.py
"""Reference outputs of the upstream koen_punctuation pipeline over results/inputs.json.

usage: /home/jiangzhuo/.cache/sokuji-punct-bench/venv-ko/bin/python parity/koen-punct.py [--backend torch|model.onnx|model.int8.onnx|model.int8-matmul.onnx]

The text output is produced by `spokentxt_punctuation_restoration.PunctuationModel.predict` (the
inference code the model card points to, spokentxt-punctuation-restoration 0.1.1 from PyPI), called
verbatim. Only its constructor is bypassed, because it cannot pin revisions: the tokenizer and model
are loaded here from the pinned snapshot with the pinned custom-code revision, then handed to it.

  --backend torch       the PyTorch model (default attention implementation, sdpa), as upstream runs it
  --backend <file>.onnx the same predict() over an onnxruntime CPU session of that graph, to separate
                        quantization effects from JS-runtime effects

Writes results/parity-koen-punct-ref-<torch|fp32|int8|int8-matmul>.json with, per row: the input, the
tokenizer ids (with <s>/</s>) and tokens, the argmax label of every position, and predict()'s output.

Rows: every ko and en row of results/inputs.json, plus one `long` row per language that joins every
`stripped` input of that language, so the JS module's 510-token windowing is exercised (upstream runs
the whole text in one pass; no corpus row comes near 510 tokens).
"""
import argparse
import hashlib
import json
import os
import sys
import time
from types import SimpleNamespace

import torch
from transformers import AutoModelForTokenClassification, AutoTokenizer

sys.path.insert(0, "/home/jiangzhuo/.claude/jobs/ac3aa5d5/tmp/spokentxt/spokentxt_punctuation_restoration-0.1.1")
from spokentxt_punctuation_restoration import PunctuationModel  # noqa: E402

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
DIR = "/home/jiangzhuo/.cache/sokuji-punct-bench/ko/koen_punctuation"
SRC = os.path.join(DIR, "src")
CODE_REV = "40ced75c3017eb27626c9d4ea981bde21a2662f4"
MODELING_SHA256 = "374670b416fcc82f081c9cd28b5fd61c2bd91bbe18eb4798fcc48a81f9c250a0"
LANGS = ["ko", "en"]


class OnnxModel:
    """Stands in for the torch model inside predict(): same call shape, logits from onnxruntime."""

    def __init__(self, path, config):
        import onnxruntime as ort

        self.sess = ort.InferenceSession(path, providers=["CPUExecutionProvider"])
        self.config = config

    def to(self, device):
        return self

    def __call__(self, input_ids, attention_mask, **_):
        logits = self.sess.run(["logits"], {"input_ids": input_ids.numpy(), "attention_mask": attention_mask.numpy()})[0]
        return SimpleNamespace(logits=torch.from_numpy(logits))


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--backend", default="torch")
    args = ap.parse_args()

    tok = AutoTokenizer.from_pretrained(SRC)
    torch_model = AutoModelForTokenClassification.from_pretrained(SRC, trust_remote_code=True, code_revision=CODE_REV).eval()
    path = sys.modules[type(torch_model).__module__].__file__
    with open(path, "rb") as fh:
        digest = hashlib.sha256(fh.read()).hexdigest()
    if digest != MODELING_SHA256:
        raise SystemExit(f"{path} is not the pinned modeling.py")
    if args.backend == "torch":
        model, stem = torch_model, "torch"
    else:
        model = OnnxModel(os.path.join(DIR, args.backend), torch_model.config)
        stem = os.path.splitext(args.backend)[0].replace("model", "", 1).lstrip(".-") or "fp32"

    pm = PunctuationModel.__new__(PunctuationModel)
    pm.tokenizer, pm.model, pm.device = tok, model, "cpu"
    normalizer = tok.backend_tokenizer.normalizer

    with open(os.path.join(ROOT, "results", "inputs.json"), encoding="utf-8") as fh:
        rows = json.load(fh)
    rows = [r for r in rows if r["lang"] in LANGS]
    for lang in LANGS:
        parts = [r["input"] for r in rows if r["lang"] == lang and r["variant"] == "stripped"]
        rows.append({"id": f"{lang}-long", "lang": lang, "variant": "long", "input": " ".join(parts)})

    out = []
    t0 = time.time()
    for r in rows:
        text = r["input"]
        enc = tok(text, return_tensors="pt")
        with torch.no_grad():
            logits = model(**enc).logits
            labels = torch.argmax(logits, dim=2).squeeze(0).tolist()
            punct = pm.predict(text)
        ids = enc["input_ids"][0].tolist()
        out.append({
            "id": r["id"], "lang": r["lang"], "variant": r["variant"], "input": text,
            "ids": ids, "tokens": tok.convert_ids_to_tokens(ids), "labels": labels, "n_tokens": len(ids),
            "normalizes_to_itself": normalizer.normalize_str(text) == text, "has_unk": tok.unk_token_id in ids,
            "punct": punct,
        })
    dst = os.path.join(ROOT, "results", f"parity-koen-punct-ref-{stem}.json")
    with open(dst, "w", encoding="utf-8") as fh:
        json.dump({"backend": args.backend, "rows": out}, fh, ensure_ascii=False, indent=1)
    print(f"wrote {dst}: {len(out)} rows in {time.time() - t0:.1f}s; max tokens {max(r['n_tokens'] for r in out)}; "
          f"rows not normalization-invariant {sum(not r['normalizes_to_itself'] for r in out)}; rows with <unk> {sum(r['has_unk'] for r in out)}")


if __name__ == "__main__":
    main()
